### PR TITLE
Fix the etcd api dependency in pkg. And Update Cobra Version to1.4.0

### PR DESCRIPTION
**Pull request by yankay**
To solve the [CVE-2020-26160](https://github.com/advisories/GHSA-w73w-5m7g-f7qc), the cobra has been upgrade to 1.2.1 before. 
And the api has been add to the pkg dependency.

But it's not necessary.   [ahrtr](https://github.com/ahrtr) advise :

“etcd already contains lots of packages and dependency relationship in-between, we shouldn't introduce unnecessary dependencies without good reason.
So I suggest to bump spf13/cobra to 1.3.0 or 1.4.0, and rollback the change on go.etcd.io/etcd/api/v3 => ./FORBIDDEN_DEPENDENCY. I would suggest to make the change on the main branch firstly, and cherry-picked to 3.5 afterwards.
” 

So the go.mod in pkg is needed to rollback ,  upgrade the corbra to 1.3.0 or 1.4.0 can solve the CVE too.

Refer to:

* https://github.com/etcd-io/etcd/pull/13797
* https://github.com/etcd-io/etcd/pull/13571

### CODE REVIEW

**Review by ahrtr — APPROVED** [review]
LGTM

Thank you!

The only concern is that the` spf13/cobra` `1.4.0` was just released 5 days ago. Usually it isn't a best practice to upgrade to a new version too soon.

**Comment by ahrtr** [comment]
cc @serathius @spzala @ptabor

**Comment by ahrtr** [comment]
> Is there any reason that we cannot just fix the api dependency without upgrading cobra further thus allow a safe cherry-pick?

We have discussed this in [pull/13797#discussion_r826516645](https://github.com/etcd-io/etcd/pull/13797#discussion_r826516645) and [pull/13797#discussion_r826596581](https://github.com/etcd-io/etcd/pull/13797#discussion_r826596581).   In short, `spf13/cobra@1.2.1` depends on `etcd` back via a long long dependency chain, so we have to introduce the unnecessary dependency `go.etcd.io/etcd/api/v3 => ../api` in pkg's go.mod. In order to remove the unnecessary dependency, we can bump `spf13/cobra` to 1.3.0 or 1.4.0.

I had a quick review on the changes between `spf13/cobra` `1.3.0` and `1.4.0`, basically there isn't much change, so it should be low risk to bump to 1.4.0.  I am open on this.  If others still have concern, we can bump to `1.3.0` for now.

**Comment by yankay** [comment]
HI @serathius . 

The main change of cobra 1.4.0, is thinned the  dependency tree. https://github.com/spf13/cobra/releases/tag/v1.4.0.
It's useful for ETCD, so the 1.4.0 version is choosed

But it's only released for 7 days . It we have concern. We can use 1.3.0 (released for 3 month) either, so that it's good to cherry pick to the ETCD 3.5.x :-)

How do you think the the change ?

**Review by serathius — APPROVED** [review]
I think it doesn't make sense to force using same cobra version for both `main` and `release-3.5` branches. We definitely don't want to have dependencies between etcd packages, so upgrading cobra to `1.4.0` on `main` is ok. 

However for `release-3.5` branch we want to make minimal changes required to fix the CVE, while not introducing the dependency. Here `1.3.0` is definitely better than `1.4.0`. 

I think we can merge this PR as it is.

**Comment by serathius** [comment]
cc @ptabor

**Comment by yankay** [comment]
> I think it doesn't make sense to force using same cobra version for both `main` and `release-3.5` branches. We definitely don't want to have dependencies between etcd packages, so upgrading cobra to `1.4.0` on `main` is ok.
> 
> However for `release-3.5` branch we want to make minimal changes required to fix the CVE, while not introducing the dependency. Here `1.3.0` is definitely better than `1.4.0`.
> 
> I think we can merge this PR as it is.

Thanks for your advise, I would create a new PR to upgrade the cobra 1.3.0 to the release 3.5.0 branch.

**Comment by yankay** [comment]
When I try to fix the release 3.5.0 branch with update cobra to 1.3.0. There are still some denpendency cycle error.

```
change the pkg/go.mod file "	github.com/spf13/cobra v1.1.3 to github.com/spf13/cobra v1.3.0 "
root@kay200:~/oss/etcd/pkg# go mod tidy
go: github.com/spf13/cobra@v1.3.0 requires
        github.com/spf13/viper@v1.10.0 requires
        go.etcd.io/etcd/api/v3@v3.5.1 (replaced by ./FORBIDDEN_DEPENDENCY): reading FORBIDDEN_DEPENDENCY/go.mod: open /root/oss/etcd/pkg/FORBIDDEN_DEPENDENCY/go.mod: no such file or directory
go: github.com/spf13/cobra@v1.3.0 requires
        github.com/spf13/viper@v1.10.0 requires
        go.etcd.io/etcd/api/v3@v3.5.1 (replaced by ./FORBIDDEN_DEPENDENCY): reading FORBIDDEN_DEPENDENCY/go.mod: open /root/oss/etcd/pkg/FORBIDDEN_DEPENDENCY/go.mod: no such file or directory
```
So I cannot successfully upgade the cobra 1.3.0 in the release 3.5.0 branch to cherry-pick the fix.

**Comment by yankay** [comment]
HI, @ptabor. 
Would you please review the PR ?

**Comment by yankay** [comment]
> Please resolve merge conflicts. cc @spzala @ahrtr for review

HI @serathius , The confict has been resolved

File: pkg/go.mod [hunk 1]
```diff
@@ -28,15 +28,10 @@ require (
...
-	go.etcd.io/etcd/api/v3 => ../api
-	go.etcd.io/etcd/client/pkg/v3 => ../client/pkg
-)
-
-// Bad imports are sometimes causing attempts to pull that code.
-// This makes the error more explicit.
 replace (
 	go.etcd.io/etcd => ./FORBIDDEN_DEPENDENCY
+	go.etcd.io/etcd/api/v3 => ./FORBIDDEN_DEPENDENCY
+	go.etcd.io/etcd/client/pkg/v3 => ../client/pkg
```
Comment (ahrtr) [inline pkg/go.mod:34]: Can you move the `go.etcd.io/etcd/client/pkg/v3 => ../client/pkg` into a separate replace block above, and keep the comment "Bad imports are sometimes cause attempts to ..." for this replace block?

Comment (yankay) [reply pkg/go.mod:34]: Good Idea.

**Comment by ahrtr** [comment]
Looks good to me.

**Comment by serathius** [comment]
Please fix conflicts.

**Comment by yankay** [comment]
HI @serathius , the confict has been fixed.  :-)

**Comment by ahrtr** [comment]
Please squash the commits. It doesn't make sense to merge multiple intermediate dev commits into the main branch.

**Review by ahrtr — CHANGES_REQUESTED** [review]
Please squash the commits

**Review by ahrtr — APPROVED** [review]

**Comment by yankay** [comment]
HI @spzala , would you please review the PR?

**Comment by serathius** [comment]
Ok, let's merge it to fix dependencies.

**Comment by yankay** [comment]
Thank you @serathius @ahrtr  very much.

**Pull request merged**